File: stripe/src/main/java/com/stripe/android/view/PaymentFlowActivity.java
package com.stripe.android.view;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.IBinder;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.viewpager.widget.ViewPager;

import com.stripe.android.CustomerSession;
import com.stripe.android.PaymentSessionData;
import com.stripe.android.R;
import com.stripe.android.model.ShippingInformation;
import com.stripe.android.model.ShippingMethod;

import java.util.List;

import static com.stripe.android.CustomerSession.EVENT_SHIPPING_INFO_SAVED;
import static com.stripe.android.CustomerSession.getInstance;
import static com.stripe.android.PaymentSession.PAYMENT_SESSION_DATA_KEY;
import static com.stripe.android.PaymentSession.TOKEN_PAYMENT_SESSION;


/**
 * Activity containing a two-part payment flow that allows users to provide a shipping address
 * as well as select a shipping method.
 */
public class PaymentFlowActivity extends StripeActivity {
    public static final String TOKEN_PAYMENT_FLOW_ACTIVITY = "PaymentFlowActivity";
    public static final String TOKEN_SHIPPING_INFO_SCREEN = "ShippingInfoScreen";
    public static final String TOKEN_SHIPPING_METHOD_SCREEN = "ShippingMethodScreen";

    private BroadcastReceiver mShippingInfoSavedBroadcastReceiver;
    private BroadcastReceiver mShippingInfoSubmittedBroadcastReceiver;
    private PaymentFlowPagerAdapter mPaymentFlowPagerAdapter;
    private ViewPager mViewPager;
    private PaymentSessionData mPaymentSessionData;
    private ShippingInformation mShippingInformationSubmitted;
    private List<ShippingMethod> mValidShippingMethods;
    private ShippingMethod mDefaultShippingMethod;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final PaymentFlowActivityStarter.Args args =
                PaymentFlowActivityStarter.Args.create(getIntent());

        final CustomerSession customerSession = getInstance();
        customerSession.addProductUsageTokenIfValid(TOKEN_PAYMENT_SESSION);
        customerSession.addProductUsageTokenIfValid(TOKEN_PAYMENT_FLOW_ACTIVITY);
        getViewStub().setLayoutResource(R.layout.activity_shipping_flow);
        getViewStub().inflate();
        mViewPager = findViewById(R.id.shipping_flow_viewpager);
        mPaymentSessionData = args.paymentSessionData;

        if (mPaymentSessionData == null) {
            throw new IllegalArgumentException(
                    "PaymentFlowActivity launched without PaymentSessionData");
        }

        mPaymentFlowPagerAdapter =
                new PaymentFlowPagerAdapter(this, args.paymentSessionConfig, customerSession);
        mViewPager.setAdapter(mPaymentFlowPagerAdapter);
        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int i, float v, int i1) {
            }

            @Override
            public void onPageSelected(int i) {
                setTitle(mViewPager.getAdapter().getPageTitle(i));
                if (mPaymentFlowPagerAdapter.getPageAt(i) == PaymentFlowPagerEnum.SHIPPING_INFO) {
                    mPaymentFlowPagerAdapter.hideShippingPage();
                }
            }

            @Override
            public void onPageScrollStateChanged(int i) {

            }
        });
        mShippingInfoSubmittedBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                final boolean isShippingInfoValid = intent.getBooleanExtra(
                        PaymentFlowExtras.EXTRA_IS_SHIPPING_INFO_VALID,
                        false);
                if (isShippingInfoValid) {
                    onShippingInfoValidated();
                    mValidShippingMethods = intent.getParcelableArrayListExtra(
                            PaymentFlowExtras.EXTRA_VALID_SHIPPING_METHODS);
                    mDefaultShippingMethod = intent
                            .getParcelableExtra(PaymentFlowExtras.EXTRA_DEFAULT_SHIPPING_METHOD);
                } else {
                    setCommunicatingProgress(false);
                    final String shippingInfoError = intent
                            .getStringExtra(PaymentFlowExtras.EXTRA_SHIPPING_INFO_ERROR);
                    if (shippingInfoError != null && !shippingInfoError.isEmpty()) {
                        showError(shippingInfoError);
                    } else {
                        showError(getString(R.string.invalid_shipping_information));
                    }
                    mShippingInformationSubmitted = null;
                }
            }
        };
        mShippingInfoSavedBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                onShippingMethodsReady(mValidShippingMethods, mDefaultShippingMethod);
                mPaymentSessionData.setShippingInformation(mShippingInformationSubmitted);
            }
        };
        setTitle(mPaymentFlowPagerAdapter.getPageTitle(mViewPager.getCurrentItem()));
    }

    @Override
    protected void onActionSave() {
        if (PaymentFlowPagerEnum.SHIPPING_INFO
                .equals(mPaymentFlowPagerAdapter.getPageAt(mViewPager.getCurrentItem()))) {
            onShippingInfoSubmitted();
        } else {
            onShippingMethodSave();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        LocalBroadcastManager.getInstance(this)
                .unregisterReceiver(mShippingInfoSubmittedBroadcastReceiver);
        LocalBroadcastManager.getInstance(this)
                .unregisterReceiver(mShippingInfoSavedBroadcastReceiver);
    }

    @Override
    protected void onResume() {
        super.onResume();
        LocalBroadcastManager.getInstance(this).registerReceiver(
                mShippingInfoSubmittedBroadcastReceiver,
                new IntentFilter(PaymentFlowExtras.EVENT_SHIPPING_INFO_PROCESSED));
        LocalBroadcastManager.getInstance(this).registerReceiver(
                mShippingInfoSavedBroadcastReceiver,
                new IntentFilter(EVENT_SHIPPING_INFO_SAVED));
    }

    private void onShippingInfoValidated() {
        CustomerSession.getInstance().setCustomerShippingInformation(
                mShippingInformationSubmitted);
    }

    private void onShippingMethodsReady(
            @NonNull List<ShippingMethod> validShippingMethods,
            @Nullable ShippingMethod defaultShippingMethod) {
        setCommunicatingProgress(false);
        mPaymentFlowPagerAdapter.setShippingMethods(validShippingMethods, defaultShippingMethod);
        mPaymentFlowPagerAdapter.setShippingInfoSaved(true);
        if (hasNextPage()) {
            mViewPager.setCurrentItem(mViewPager.getCurrentItem() + 1);
        } else {
            mPaymentSessionData.setShippingInformation(mShippingInformationSubmitted);
            setResult(RESULT_OK,
                    new Intent().putExtra(PAYMENT_SESSION_DATA_KEY, mPaymentSessionData));
            finish();
        }
    }

    private void onShippingInfoSubmitted() {
        hideKeyboard();

        final ShippingInfoWidget shippingInfoWidget = findViewById(R.id.shipping_info_widget);
        final ShippingInformation shippingInformation = shippingInfoWidget.getShippingInformation();
        if (shippingInformation != null) {
            mShippingInformationSubmitted = shippingInformation;
            setCommunicatingProgress(true);
            broadcastShippingInfoSubmitted(shippingInformation);
        }
    }

    private void hideKeyboard() {
        final InputMethodManager inputMethodManager =
                (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (inputMethodManager.isAcceptingText()) {
            final View currentFocus = getCurrentFocus();
            final IBinder windowToken = currentFocus != null ? currentFocus.getWindowToken() : null;
            inputMethodManager.hideSoftInputFromWindow(windowToken, 0);
        }
    }

    private void broadcastShippingInfoSubmitted(ShippingInformation shippingInformation) {
        LocalBroadcastManager
                .getInstance(this)
                .sendBroadcast(
                        new Intent(PaymentFlowExtras.EVENT_SHIPPING_INFO_SUBMITTED)
                                .putExtra(PaymentFlowExtras.EXTRA_SHIPPING_INFO_DATA,
                                        shippingInformation));
    }

    private boolean hasNextPage() {
        return mViewPager.getCurrentItem() + 1 < mPaymentFlowPagerAdapter.getCount();
    }

    private boolean hasPreviousPage() {
        int currentPageIndex = mViewPager.getCurrentItem();
        return currentPageIndex != 0;
    }

    private void onShippingMethodSave() {
        final SelectShippingMethodWidget selectShippingMethodWidget =
                findViewById(R.id.select_shipping_method_widget);
        final ShippingMethod shippingMethod = selectShippingMethodWidget
                .getSelectedShippingMethod();
        mPaymentSessionData.setShippingMethod(shippingMethod);
        setResult(RESULT_OK,
                new Intent().putExtra(PAYMENT_SESSION_DATA_KEY, mPaymentSessionData));
        finish();
    }

    @Override
    public void onBackPressed() {
        if (hasPreviousPage()) {
            mViewPager.setCurrentItem(mViewPager.getCurrentItem() - 1);
            return;
        }
        super.onBackPressed();
    }
}

File: stripe/src/main/java/com/stripe/android/view/PaymentFlowActivity.kt
package com.stripe.android.view

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.viewpager.widget.ViewPager
import com.stripe.android.CustomerSession
import com.stripe.android.PaymentSession.Companion.EXTRA_PAYMENT_SESSION_DATA
import com.stripe.android.PaymentSessionConfig
import com.stripe.android.PaymentSessionData
import com.stripe.android.R
import com.stripe.android.databinding.PaymentFlowActivityBinding
import com.stripe.android.model.ShippingInformation
import com.stripe.android.model.ShippingMethod

/**
 * Activity containing a two-part payment flow that allows users to provide a shipping address
 * as well as select a shipping method.
 */
class PaymentFlowActivity : StripeActivity() {

    private val viewBinding: PaymentFlowActivityBinding by lazy {
        viewStub.layoutResource = R.layout.payment_flow_activity
        val root = viewStub.inflate() as ViewGroup
        PaymentFlowActivityBinding.bind(root)
    }

    private val viewPager: PaymentFlowViewPager by lazy {
        viewBinding.shippingFlowViewpager
    }

    private val customerSession: CustomerSession by lazy {
        CustomerSession.getInstance()
    }
    private val args: PaymentFlowActivityStarter.Args by lazy {
        PaymentFlowActivityStarter.Args.create(intent)
    }
    private val paymentSessionConfig: PaymentSessionConfig by lazy {
        args.paymentSessionConfig
    }
    private val viewModel: PaymentFlowViewModel by lazy {
        ViewModelProvider(
            this,
            PaymentFlowViewModel.Factory(customerSession, args.paymentSessionData)
        )[PaymentFlowViewModel::class.java]
    }

    private val paymentFlowPagerAdapter: PaymentFlowPagerAdapter by lazy {
        PaymentFlowPagerAdapter(
            this,
            paymentSessionConfig,
            paymentSessionConfig.allowedShippingCountryCodes
        ) {
            viewModel.selectedShippingMethod = it
        }
    }

    private val keyboardController: KeyboardController by lazy {
        KeyboardController(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val args = PaymentFlowActivityStarter.Args.create(intent)
        args.windowFlags?.let { window.addFlags(it) }

        val shippingInformation =
            viewModel.submittedShippingInfo
                ?: paymentSessionConfig.prepopulatedShippingInfo

        paymentFlowPagerAdapter.shippingMethods = viewModel.shippingMethods
        paymentFlowPagerAdapter.isShippingInfoSubmitted = viewModel.isShippingInfoSubmitted
        paymentFlowPagerAdapter.shippingInformation = shippingInformation
        paymentFlowPagerAdapter.selectedShippingMethod = viewModel.selectedShippingMethod

        viewPager.adapter = paymentFlowPagerAdapter
        viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(i: Int, v: Float, i1: Int) {}

            override fun onPageSelected(i: Int) {
                title = paymentFlowPagerAdapter.getPageTitle(i)
                if (paymentFlowPagerAdapter.getPageAt(i) === PaymentFlowPage.ShippingInfo) {
                    viewModel.isShippingInfoSubmitted = false
                    paymentFlowPagerAdapter.isShippingInfoSubmitted = false
                }
            }

            override fun onPageScrollStateChanged(i: Int) {
            }
        })

        viewPager.currentItem = viewModel.currentPage
        title = paymentFlowPagerAdapter.getPageTitle(viewPager.currentItem)
    }

    public override fun onActionSave() {
        if (PaymentFlowPage.ShippingInfo ==
            paymentFlowPagerAdapter.getPageAt(viewPager.currentItem)) {
            onShippingInfoSubmitted()
        } else {
            onShippingMethodSave()
        }
    }

    @JvmSynthetic
    internal fun onShippingInfoSaved(
        shippingInformation: ShippingInformation?,
        shippingMethods: List<ShippingMethod> = emptyList()
    ) {
        onShippingMethodsReady(shippingMethods)
        viewModel.paymentSessionData = viewModel.paymentSessionData.copy(
            shippingInformation = shippingInformation
        )
    }

    private fun onShippingInfoValidated(shippingMethods: List<ShippingMethod>) {
        viewModel.paymentSessionData.shippingInformation?.let { shippingInfo ->
            viewModel.saveCustomerShippingInformation(shippingInfo)
                .observe(this, Observer {
                    when (it) {
                        is PaymentFlowViewModel.SaveCustomerShippingInfoResult.Success -> {
                            onShippingInfoSaved(
                                it.customer.shippingInformation,
                                shippingMethods
                            )
                        }
                        is PaymentFlowViewModel.SaveCustomerShippingInfoResult.Error -> {
                            showError(it.errorMessage)
                        }
                    }
                })
        }
    }

    private fun onShippingMethodsReady(
        shippingMethods: List<ShippingMethod>
    ) {
        isProgressBarVisible = false
        paymentFlowPagerAdapter.shippingMethods = shippingMethods
        paymentFlowPagerAdapter.isShippingInfoSubmitted = true

        if (hasNextPage()) {
            viewModel.currentPage += 1
            viewPager.currentItem = viewModel.currentPage
        } else {
            finishWithData(viewModel.paymentSessionData)
        }
    }

    private fun onShippingInfoSubmitted() {
        keyboardController.hide()

        shippingInfo?.let { shippingInfo ->
            viewModel.paymentSessionData = viewModel.paymentSessionData.copy(
                shippingInformation = shippingInfo
            )
            isProgressBarVisible = true

            validateShippingInformation(
                paymentSessionConfig.shippingInformationValidator,
                paymentSessionConfig.shippingMethodsFactory,
                shippingInfo
            )
        }
    }

    private val shippingInfo: ShippingInformation?
        get() {
            return viewPager
                .findViewById<ShippingInfoWidget>(R.id.shipping_info_widget)
                .shippingInformation
        }

    private fun hasNextPage(): Boolean {
        return viewPager.currentItem + 1 < paymentFlowPagerAdapter.count
    }

    private fun hasPreviousPage(): Boolean {
        return viewPager.currentItem != 0
    }

    private fun onShippingMethodSave() {
        val selectedShippingMethod =
            viewPager
                .findViewById<SelectShippingMethodWidget>(R.id.select_shipping_method_widget)
                .selectedShippingMethod
        finishWithData(viewModel.paymentSessionData.copy(
            shippingMethod = selectedShippingMethod
        ))
    }

    private fun validateShippingInformation(
        shippingInfoValidator: PaymentSessionConfig.ShippingInformationValidator,
        shippingMethodsFactory: PaymentSessionConfig.ShippingMethodsFactory?,
        shippingInformation: ShippingInformation
    ) {
        viewModel.validateShippingInformation(
            shippingInfoValidator,
            shippingMethodsFactory,
            shippingInformation
        ).observe(this, Observer {
            when (it) {
                is PaymentFlowViewModel.ValidateShippingInfoResult.Success -> {
                    // show shipping methods screen
                    onShippingInfoValidated(it.shippingMethods)
                }
                is PaymentFlowViewModel.ValidateShippingInfoResult.Error -> {
                    // show error on current screen
                    onShippingInfoError(it.errorMessage)
                }
            }
        })
    }

    private fun onShippingInfoError(errorMessage: String?) {
        isProgressBarVisible = false
        if (!errorMessage.isNullOrEmpty()) {
            showError(errorMessage)
        } else {
            showError(getString(R.string.invalid_shipping_information))
        }
        viewModel.paymentSessionData = viewModel.paymentSessionData.copy(
            shippingInformation = null
        )
    }

    private fun finishWithData(paymentSessionData: PaymentSessionData) {
        setResult(Activity.RESULT_OK,
            Intent().putExtra(EXTRA_PAYMENT_SESSION_DATA, paymentSessionData)
        )
        finish()
    }

    override fun onBackPressed() {
        if (hasPreviousPage()) {
            viewModel.currentPage -= 1
            viewPager.currentItem = viewModel.currentPage
        } else {
            super.onBackPressed()
        }
    }

    internal companion object {
        internal const val PRODUCT_TOKEN: String = "PaymentFlowActivity"
    }
}
